src/rgw/rgw_data_sync.cc

```diff
@@ -3387,193 +4685,268 @@ int RGWRunBucketSyncCoroutine::operate()
+int RGWBucketPipeSyncStatusManager::read_sync_status()
 {
   list<RGWCoroutinesStack *> stacks;
 
-  for (map<int, RGWRemoteBucketLog *>::iterator iter = source_logs.begin(); iter != source_logs.end(); ++iter) {
+  for (auto& mgr : source_mgrs) {
     RGWCoroutinesStack *stack = new RGWCoroutinesStack(store->ctx(), &cr_mgr);
-    RGWRemoteBucketLog *l = iter->second;
-    stack->call(l->read_sync_status_cr(&sync_status[iter->first]));
+    for (int i = 0; i < mgr->num_pipes(); ++i) {
+      stack->call(mgr->read_sync_status_cr(i, &sync_status[i]));
+    }
 
     stacks.push_back(stack);
   }
 
   int ret = cr_mgr.run(stacks);
   if (ret < 0) {
     ldpp_dout(this, 0) << "ERROR: failed to read sync status for "
-        << bucket_str{bucket} << dendl;
+        << bucket_str{dest_bucket} << dendl;
     return ret;
   }
 
   return 0;
 }
 
-int RGWBucketSyncStatusManager::run()
+int RGWBucketPipeSyncStatusManager::run()
 {
   list<RGWCoroutinesStack *> stacks;
 
-  for (map<int, RGWRemoteBucketLog *>::iterator iter = source_logs.begin(); iter != source_logs.end(); ++iter) {
+  for (auto& mgr : source_mgrs) {
     RGWCoroutinesStack *stack = new RGWCoroutinesStack(store->ctx(), &cr_mgr);
-    RGWRemoteBucketLog *l = iter->second;
-    stack->call(l->run_sync_cr());
+    for (int i = 0; i < mgr->num_pipes(); ++i) {
+      stack->call(mgr->run_sync_cr(i));
+    }
 
     stacks.push_back(stack);
   }
 
   int ret = cr_mgr.run(stacks);
   if (ret < 0) {
     ldpp_dout(this, 0) << "ERROR: failed to read sync status for "
-        << bucket_str{bucket} << dendl;
+        << bucket_str{dest_bucket} << dendl;
     return ret;
   }
 
   return 0;
 }
 
-unsigned RGWBucketSyncStatusManager::get_subsys() const
+unsigned RGWBucketPipeSyncStatusManager::get_subsys() const
 {
   return dout_subsys;
 }
 
-std::ostream& RGWBucketSyncStatusManager::gen_prefix(std::ostream& out) const
+std::ostream& RGWBucketPipeSyncStatusManager::gen_prefix(std::ostream& out) const
 {
-  auto zone = std::string_view{source_zone};
+  auto zone = std::string_view{source_zone.value_or(rgw_zone_id("*")).id};
   return out << "bucket sync zone:" << zone.substr(0, 8)
-      << " bucket:" << bucket.name << ' ';
+    << " bucket:" << dest_bucket << ' ';
 }
 
-string RGWBucketSyncStatusManager::status_oid(const string& source_zone,
-                                              const rgw_bucket_shard& bs)
+string RGWBucketPipeSyncStatusManager::status_oid(const rgw_zone_id& source_zone,
+                                              const rgw_bucket_sync_pair_info& sync_pair)
 {
-  return bucket_status_oid_prefix + "." + source_zone + ":" + bs.get_key();
+  if (sync_pair.source_bs == sync_pair.dest_bs) {
+    return bucket_status_oid_prefix + "." + source_zone.id + ":" + sync_pair.dest_bs.get_key();
+  } else {
+    return bucket_status_oid_prefix + "." + source_zone.id + ":" + sync_pair.dest_bs.get_key() + ":" + sync_pair.source_bs.get_key();
```

Comment:
conversely, if a given source bucket is only configured to replicate to some other destination bucket, it would only write sync status to the object named on line 4832 here

since bilog trimming is only requesting the bucket sync status from the object named on line 4830, it would see an empty sync status, and would never attempt to trim its bilog

Reply:
@cbodley addressed this here:
https://github.com/yehudasa/ceph/tree/wip-rgw-sync-bucket-6-review

 - new option for fetching aggregated minimum sync status from target buckets in a specific zone
 - trimmer fetch status from relevant zones

Reply:
thanks @yehudasa, can you please add those commits to the pr branch so i can review here?

Reply:
@cbodley pushed